styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -91,72 +104,110 @@ void tick() {
-      LOG.warn("workflow not found for backfill, skipping rest of triggers: {}", backfill);
-      final BackfillBuilder builder = backfill.builder();
-      builder.halted(true);
-      storeBackfill(builder.build());
+      LOG.debug("workflow not found for backfill {}, halt it.", backfill);
+      storeBackfill(backfill.builder().halted(true).build());
       return;
     }
 
     final Workflow workflow = workflowOpt.get();
 
+    // this is best effort because querying active states is not strongly consistent so the
+    // initial remaining capacity may already be wrong
     final int remainingCapacity =
-        backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+        backfill.concurrency() - stateManager.activeStatesByTriggerId(backfill.id()).size();
 
-    Instant partition = backfill.nextTrigger();
+    if (remainingCapacity < 1) {
+      LOG.debug("No capacity left for backfill {}", backfill);
+      return;
+    }
 
-    for (int i = 0; i < remainingCapacity && partition.isBefore(backfill.end()); i++) {
+    final Instant initialNextTrigger = backfill.nextTrigger();
+    while (true) {
       try {
-        final CompletableFuture<Void> processed = triggerListener
-            .event(workflow, Trigger.backfill(backfill.id()), partition)
-            .toCompletableFuture();
-        // Wait for the trigger execution to complete before proceeding to the next partition
-        processed.get();
-      } catch (AlreadyInitializedException ignored) {
-        // nop
-      } catch (Throwable e) {
-        LOG.warn("failed to trigger backfill for state [{}]: {}", partition, backfill, e);
-        return;
+        if (!storage.runInTransaction(tx ->
+            triggerNextPartitionAndProgress(tx, backfill.id(), workflow,
+                initialNextTrigger, remainingCapacity))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("Transaction failure when working on backfill {}", backfill, e);
+        break;
       }
+    }
+  }
 
-      partition = nextInstant(partition, backfill.schedule());
-      storeBackfill(backfill.builder()
-                        .nextTrigger(partition)
-                        .build());
+  private boolean triggerNextPartitionAndProgress(StorageTransaction tx,
+                                                  String id,
+                                                  Workflow workflow,
+                                                  Instant initialNextTrigger,
+                                                  int remainingCapacity) {
+    final Backfill momentBackfill = tx.backfill(id).orElseThrow(RuntimeException::new);
+    final Instant momentNextTrigger = momentBackfill.nextTrigger();
+
+    if (numberOfInstants(momentNextTrigger, initialNextTrigger,
+        momentBackfill.schedule()) >= remainingCapacity) {
+      LOG.debug("Capacity reached for backfill {}", momentBackfill);
```

Comment:
The `lastInstant` and `firstInstant` were both inclusive, so the comparison doesn't exactly make sense. If `remainingCapacity` is 1 and `momentNextTrigger` equals `initialNextTrigger`, then I expect this WFI to be triggered, but if `numberOfInstants()` behaves as documented then we are comparing `1 >= 1` and saying no to it...?